File: solutions/Permutations/Permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
  const ans = [];
  const dfs = (vec, pos) => {
    if (pos === vec.length) {
      ans.push(vec.slice());
    }
    for (let i = pos; i < vec.length; i++) {
      [vec[pos], vec[i]] = [vec[i], vec[pos]];
      dfs(vec, pos + 1);
      [vec[i], vec[pos]] = [vec[pos], vec[i]];
    }
  };
  dfs(nums.slice(), 0);
  return ans;
};

File: solutions/Permutations/Permutations.cpp

// class Solution {
// public:
//     void dfs(vector<vector<int> > &ans,vector<int> &num,int pos)
//     {
//         if(pos==num.size())
//             ans.push_back(num);
//         for(int i=pos;i<num.size();i++)
//         {
//             swap(num[pos],num[i]);
//             dfs(ans,num,pos+1);
//             swap(num[pos],num[i]);
//         }
//     }
//     vector<vector<int> > permute(vector<int> &num) {
//         vector<vector<int> >ans;
//         dfs(ans,num,0);
//         return ans;
//     }
// };

// 2019-07-20
// O(n!)
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        dfs(nums, 0,  ans);
        return ans;
    }
private:
    void dfs(vector<int>& nums, int index, vector<vector<int>>& ans) {
        if (index == nums.size())
        {
            ans.push_back(nums);
            return;
        }
        for(int i=index;i<nums.size();i++)
        {
            swap(nums[i], nums[index]);
            dfs(nums, index + 1, ans);
            swap(nums[i], nums[index]);
        }
    }
};
 